### PR TITLE
implement exit

### PR DESCRIPTION
**Pull request by RalfJung**

### CODE REVIEW

**Comment by bjorn3** [comment]
Could you please just pop the whole stack or something like that. This will make pritoda exit when `exit` is called, which makes it impossible to for example show a flamegraph of the program execution.

**Comment by RalfJung** [comment]
That wouldn't be right... in Miri we do want the process to quit with the status code the interpreted program would also use.

I suppose I could make an `ExitError` that's not actually an error, and that gets caught in `eval_main`. That requires a rustc change though.

**Comment by bjorn3** [comment]
Maybe return `MachineError(format!("exit code: {}", code))` and catch it in `eval_main`. That way priroda will automatically show that the process exited that way, but doesn't exit.

**Comment by RalfJung** [comment]
I refuse to write code that catches an error and interprets the string to figure out which error it is.^^

I'll add a new error variant instead.

**Comment by bjorn3** [comment]
> I'll add a new error variant instead.

+1

**Comment by RalfJung** [comment]
Here you go: https://github.com/rust-lang/rust/pull/60116

**Comment by RalfJung** [comment]
@bjorn3 does this work for you?

File: src/fn_call.rs [hunk 1]
```diff
@@ -149,11 +149,19 @@ pub trait EvalContextExt<'a, 'mir, 'tcx: 'a + 'mir>: crate::MiriEvalContextExt<'
...
         match link_name {
             "__rust_start_panic" | "panic_impl" => {
                 return err!(MachineError("the evaluated program panicked".to_string()));
             }
+            "exit" => {
+                let code = this.read_scalar(args[0])?.to_i32()?;
+                return err!(Exit(code));
+            }
+            "ExitProcess" => {
+                let code = this.read_scalar(args[0])?.to_u32()?;
```
Comment (oli-obk) [inline src/fn_call.rs]: even if `ExitProcess` has a `u32` return type, if we're just casting to `i32` immediately we can use `to_i32` here and deduplicate the code between `exit` and `ExitProcess`

File: tests/run-pass/exit.rs [hunk 2]
```diff
@@ -0,0 +1,3 @@
+fn main() {
+    std::process::exit(0)
```
Comment (oli-obk) [inline tests/run-pass/exit.rs]: Also add a run-fail test?

File: src/fn_call.rs (hunk 1 above)
Comment (RalfJung) [reply src/fn_call.rs]: Done.

File: tests/run-pass/exit.rs (hunk 2 above)
Comment (RalfJung) [reply tests/run-pass/exit.rs]: We only have `compile-fail`... and what would the error pattern look like?

We don't dispatch based on the actual exit code at all, so this seems like a fine test to me.

Comment (oli-obk) [reply tests/run-pass/exit.rs]: hmm... not sure if compiletest-rs has the comment flags for requiring specific exit codes like rustc's compiletest suite has.

So yea, seems fine this way

**Pull request merged**